src/dfx/src/lib/builders/custom.rs

```diff
@@ -150,6 +150,14 @@ impl CanisterBuilder for CustomBuilder {
             add_candid_service_metadata = false;
         }
 
+        // Custom canister may have WASM gzipped
+        if config.shrink_after_build
+            && wasm.extension().and_then(std::ffi::OsStr::to_str) == Some("wasm")
+        {
+            info!(self.logger, "Shrink WASM module size.");
+            super::shrink_wasm(&wasm)?;
+        }
```

Comment:
Instead of looking at the file extension, we can inspect the file header.  This information is already available in this function, immediately above, lines 145-150.